### PR TITLE
[Feature] - templates for job foms

### PR DESCRIPTION
**Pull request by ArvApu**

### CODE REVIEW

**Review by erteriuke — DISMISSED** [review]
For me it looks great, maybe someone else should also do a review? :D 💯

File: api/app/Http/Controllers/JobFormController.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+     * @throws \Illuminate\Validation\ValidationException
+     */
+    public function store(Request $request)
+    {
+        $validators = $this->getValidators();
+        $attributes = $this->validate($request, $validators);
+
+        JobForm::create($attributes + ['user_id' => auth()->user()->id]);
+
+        return redirect()->back()->with('success', 'Job form created');
```
Comment (EdotJ) [inline api/app/Http/Controllers/JobFormController.php]: I would prefer it redirected to job forms index instead of the same create page.

File: api/app/Http/Controllers/JobFormController.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        if(!auth()->user()->isJobFormOwner($jobForm->id))
+            return redirect()->back()->withErrors('Forbidden');
+
+        $jobForm->delete();
+        return redirect()->back()->with('success', 'Job form was deleted');
+    }
+
+    /**
+     * Return array of fields that ahs to be validated and their rules
```
Comment (EdotJ) [inline api/app/Http/Controllers/JobFormController.php]: has*

File: api/app/Http/Controllers/JobFormController.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * Show the form for creating a new resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function create()
+    {
+        return view('job-form.create',[
+            'fields' => $this->getTableColumns('job_forms')->reject(function ($column) {
```
Comment (EdotJ) [inline api/app/Http/Controllers/JobFormController.php]: Couldn't this be created in the constructor? Wouldn't need to call the function twice, less cluttery. The database structure shouldn't change mid-call.

File: api/resources/views/layouts/app.blade.php [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
     <!-- Scripts -->
     <script src="{{ asset('js/app.js') }}" defer></script>
+    <script src="{{ asset('js/main.js') }}" defer></script>
```
Comment (EdotJ) [inline api/resources/views/layouts/app.blade.php:14]: Why is this js imported here?

File: api/resources/views/job-form/show.blade.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
+@extends('layouts.app')
+
+@section('content')
+
+    <div class="container mt-4">
+        @include('errors')
+        <form action="{{ route('filled-forms.store') }}" method="post">
```
Comment (EdotJ) [inline api/resources/views/job-form/show.blade.php:7]: The inputs reset after failed attempt to submit. Fix that, terrible UX.

File: api/resources/views/job-form/index.blade.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        <div>
+            <form class="form-inline d-flex justify-content-between align-items-center">
+                <label class="sr-only" for="form_name">Form name</label>
+                <input type="text" class="form-control mb-2 mr-sm-2" id="form_name" name="form_name" placeholder="Form name">
+
+                <label class="sr-only" for="creator">Creator</label>
+                <input type="text" class="form-control mb-2 mr-sm-2" id="creator" name="creator" placeholder="Creator">
+
+                <label class="sr-only" for="date_from">Date from</label>
+                <input type="text" class="form-control mb-2 mr-sm-2" id="date_from" name="date_from" placeholder="Created after date">
```
Comment (EdotJ) [inline api/resources/views/job-form/index.blade.php]: Date picker for date inputs please.

File: api/resources/views/job-form/index.blade.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+                <label class="sr-only" for="date_to">Name</label>
+                <input type="text" class="form-control mb-2 mr-sm-2" id="date_to" name="date_to" placeholder="Created before date">
```
Comment (EdotJ) [inline api/resources/views/job-form/index.blade.php]: ditto

File: api/resources/views/job-form/edit.blade.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            @method('put')
+
+            <div class="form-group">
+                <label for="name">Form Name</label>
+                <input type="text" class="form-control form-control" id="name" name="name" placeholder="Enter form name" value="{{$jobForm->name}}">
+            </div>
+
+            @foreach($fields as $field)
+                <div class="form-group">
+                    <label style="text-transform: capitalize" for="select-{{$field}}">{{ implode(" ", explode('_', $field)) }} field</label>
```
Comment (EdotJ) [inline api/resources/views/job-form/edit.blade.php]: Could've used bootstrap's text-capitalize class instead of an inline style tag.

File: api/resources/views/job-form/edit.blade.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                    <select class="custom-select custom-select-lg" name="{{$field}}">
+                        <option disabled>Select status for {{ implode(" ", explode('_', $field)) }} field</option>
```
Comment (EdotJ) [inline api/resources/views/job-form/edit.blade.php:20]: I'd prefer there was a default choice, either unused or optional, instead of having to manually choose everything every time. Plus, fields reset after failed submit here too.

File: api/resources/views/job-form/create.blade.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            @csrf
+
+            <div class="form-group">
+                <label for="name">Form Name</label>
+                <input type="text" class="form-control form-control" id="name" name="name" placeholder="Enter form name">
+            </div>
+
+            @foreach($fields as $field)
+                <div class="form-group">
+                    <label style="text-transform: capitalize" for="select-{{$field}}">{{ implode(" ", explode('_', $field)) }} field</label>
```
Comment (EdotJ) [inline api/resources/views/job-form/create.blade.php]: Ditto with class text-capitalize

File: api/resources/views/job-form/create.blade.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+                    <select id="select-{{$field}}" class="custom-select custom-select-lg" name="{{$field}}">
+                        <option selected disabled>Select status for {{ implode(" ", explode('_', $field)) }} field</option>
```
Comment (EdotJ) [inline api/resources/views/job-form/create.blade.php]: ditto from edit.blade.php

File: api/resources/views/filled-form/show.blade.php [hunk 12]
```diff
@@ -0,0 +1,17 @@
+@extends('layouts.app')
+
+@section('content')
+
+    <div class="container mt-4">
+        <h1>Filled {{ $form->jobForm->name }} form</h1>
```
Comment (EdotJ) [inline api/resources/views/filled-form/show.blade.php]: maybe put this in a card instead of a weird card-like structure? 
```
<div class="card">
  <div class="card-header">
    <h1> Header </h1>
  </div>
  <div class="card-body">
    {{content}
  </div>
</div>
```

File: api/resources/views/filled-form/edit.blade.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+    <div class="container mt-4">
+        @include('errors')
+        <form action="{{ route('filled-forms.store') }}" method="post">
+            @csrf
+            @foreach($fields as $field => $status)
+                <div class="form-group">
+                    @if($statuses->get($status) !== 'unused' )
+                        <label style="text-transform: capitalize" for="{{$field}}">{{ implode(" ", explode('_', $field)) }}</label>
+                        <input type="text" class="form-control form-control" id="{{$field}}" name="{{$field}}"
```
Comment (EdotJ) [inline api/resources/views/filled-form/edit.blade.php:13]: Would there be a way to validate these fields? Right now I can enter REEE as a phone number, which is not really good. Same with when filling a form.

File: api/public/js/main.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+function deleteAction (endpoint, token) {
```
Comment (EdotJ) [inline api/public/js/main.js:1]: Could you enlighten me why this is used when deleting? Is it to not show an id, not allow other users to try and delete it or something? This could then go to app.js instead of a public folder, or is that not feasible?

File: api/database/migrations/2019_11_09_123934_rename_column_for_job_forms_table.php [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+class RenameColumnForJobFormsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('job_forms', function(Blueprint $table) {
```
Comment (EdotJ) [inline api/database/migrations/2019_11_09_123934_rename_column_for_job_forms_table.php:16]: I would've prefered to have both tables changed in one migration but this works too.

File: api/resources/views/filled-form/edit.blade.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+                               placeholder="Enter {{ implode(" ", explode('_', $field)) }}"
+                               value = "{{ $filledForm->$field }}">
+                    @endif
+                </div>
+            @endforeach
+
+            <input hidden style="display: none" type="text" name="formId" value="{{ $jobFormId }}">
```
Comment (EdotJ) [inline api/resources/views/filled-form/edit.blade.php]: Why is this hidden AND display:none? if that's really necessary, class="d-none" instead of an inline style, perhaps?

File: api/app/User.php [hunk 17]
```diff
@@ -36,4 +37,19 @@ class User extends Authenticatable
...
     ];
+
+    public function jobForms()
+    {
+        return $this->hasMany(JobForm::class);
+    }
+
+    public function isJobFormOwner($id)
+    {
+        return in_array($id, $this->jobForms()->pluck('id')->all());
```
Comment (EdotJ) [inline api/app/User.php:47]: Why was this needed? Why not just check auth'ed user id with a form's user_id? Unless this is returning if a user has forms filled, which I doubt.

File: api/app/User.php [hunk 18]
```diff
@@ -5,6 +5,7 @@
 use Illuminate\Contracts\Auth\MustVerifyEmail;
 use Illuminate\Foundation\Auth\User as Authenticatable;
 use Illuminate\Notifications\Notifiable;
+use Illuminate\Queue\Jobs\Job;
```
Comment (EdotJ) [inline api/app/User.php]: unused import.

**Review by EdotJ — CHANGES_REQUESTED** [review]
Some things I can let slide although some should definitely changed, especially the form resetting.

File: api/resources/views/layouts/app.blade.php (hunk 4 above)
Comment (ArvApu) [reply api/resources/views/layouts/app.blade.php:14]: cuz we use its code?

File: api/resources/views/job-form/show.blade.php (hunk 5 above)
Comment (ArvApu) [reply api/resources/views/job-form/show.blade.php:7]: Okay

File: api/resources/views/job-form/edit.blade.php (hunk 8 above)
Comment (ArvApu) [reply api/resources/views/job-form/edit.blade.php]: yeahh

File: api/resources/views/job-form/edit.blade.php (hunk 9 above)
Comment (ArvApu) [reply api/resources/views/job-form/edit.blade.php:20]: No default choice and will fix that reset

File: api/resources/views/filled-form/edit.blade.php (hunk 13 above)
Comment (ArvApu) [reply api/resources/views/filled-form/edit.blade.php:13]: It is jus strings which we should realy care what is inputed, because for egzample we would need regular exp for numbers, lets just leave like rel life paper version of form, you can write whatever you like in it

File: api/public/js/main.js (hunk 14 above)
Comment (ArvApu) [reply api/public/js/main.js:1]: Iskeliau trinima i js funckiaj kad nereiktu hidden formu kurti o paskui jas submitinti.

File: api/resources/views/filled-form/edit.blade.php (hunk 16 above)
Comment (ArvApu) [reply api/resources/views/filled-form/edit.blade.php]: double hidde motherfu**er :DD jk , going to fix it

File: api/app/User.php (hunk 17 above)
Comment (ArvApu) [reply api/app/User.php:47]: user can have many job forms. We get list of all his job forms. We check if given job form is in the list. Your option should be implemented in JobForm class. And i wanted to check if user is owner and if job form is owned by, because after admi nrole will be added you just can add || $this->isAdmin(); Ofc you can do similary in JobsForm. However, i did not implemnted filled forms owner and my idea was to put both checks under one method.  This will i look into a bit later.

File: api/app/User.php (hunk 18 above)
Comment (ArvApu) [reply api/app/User.php]: WTF. will fix

File: api/resources/views/filled-form/show.blade.php (hunk 12 above)
Comment (ArvApu) [reply api/resources/views/filled-form/show.blade.php]: If i liek the style i would take the card approach

File: api/resources/views/job-form/create.blade.php (hunk 11 above)
Comment (ArvApu) [reply api/resources/views/job-form/create.blade.php]: Ok.

File: api/resources/views/job-form/create.blade.php (hunk 10 above)
Comment (ArvApu) [reply api/resources/views/job-form/create.blade.php]: OK.

File: api/resources/views/job-form/index.blade.php (hunk 6 above)
Comment (ArvApu) [reply api/resources/views/job-form/index.blade.php]: Great

File: api/app/Http/Controllers/JobFormController.php (hunk 3 above)
Comment (ArvApu) [reply api/app/Http/Controllers/JobFormController.php]: Great idea, actually, improves readability. However does not make any more difference.

File: api/app/Http/Controllers/JobFormController.php (hunk 1 above)
Comment (ArvApu) [reply api/app/Http/Controllers/JobFormController.php]: Done

File: api/app/Http/Controllers/JobFormController.php (hunk 2 above)
Comment (ArvApu) [reply api/app/Http/Controllers/JobFormController.php]: done

File: api/app/Http/Controllers/JobFormController.php (hunk 3 above)
Comment (ArvApu) [reply api/app/Http/Controllers/JobFormController.php]: done

File: api/resources/views/layouts/app.blade.php (hunk 4 above)
Comment (ArvApu) [reply api/resources/views/layouts/app.blade.php:14]: No fix needed

File: api/resources/views/job-form/show.blade.php (hunk 5 above)
Comment (ArvApu) [reply api/resources/views/job-form/show.blade.php:7]: Done

File: api/public/js/main.js (hunk 14 above)
Comment (ArvApu) [reply api/public/js/main.js:1]: no fix needed, app.js is better not used for modifactions beaces it has prewriten code

File: api/database/migrations/2019_11_09_123934_rename_column_for_job_forms_table.php (hunk 15 above)
Comment (ArvApu) [reply api/database/migrations/2019_11_09_123934_rename_column_for_job_forms_table.php:16]: no fiex needed

File: api/resources/views/filled-form/edit.blade.php (hunk 16 above)
Comment (ArvApu) [reply api/resources/views/filled-form/edit.blade.php]: done

File: api/app/User.php (hunk 17 above)
Comment (ArvApu) [reply api/app/User.php:47]: will fix latter if needed, when implementing other checks

File: api/app/User.php (hunk 18 above)
Comment (ArvApu) [reply api/app/User.php]: done

File: api/resources/views/job-form/index.blade.php (hunk 6 above)
Comment (ArvApu) [reply api/resources/views/job-form/index.blade.php]: done

File: api/resources/views/job-form/index.blade.php (hunk 7 above)
Comment (ArvApu) [reply api/resources/views/job-form/index.blade.php]: done

File: api/resources/views/job-form/edit.blade.php (hunk 8 above)
Comment (ArvApu) [reply api/resources/views/job-form/edit.blade.php]: done

File: api/resources/views/job-form/edit.blade.php (hunk 9 above)
Comment (ArvApu) [reply api/resources/views/job-form/edit.blade.php:20]: It is an edit so, it has predifend values, no reset to fix in here

File: api/resources/views/job-form/create.blade.php (hunk 10 above)
Comment (ArvApu) [reply api/resources/views/job-form/create.blade.php]: done

File: api/resources/views/job-form/create.blade.php (hunk 11 above)
Comment (ArvApu) [reply api/resources/views/job-form/create.blade.php]: ^

File: api/resources/views/filled-form/show.blade.php (hunk 12 above)
Comment (ArvApu) [reply api/resources/views/filled-form/show.blade.php]: done

File: api/public/js/main.js (hunk 14 above)
Comment (EdotJ) [reply api/public/js/main.js:1]: Fair point for hidden forms. But I still don't like it being in the public folder. 
As far as I know, app.js has only `require('./bootstrap');`. You're looking at public/js/app.js. Laravel stores pre-compiled javascript files (the files you write code in) in resources/js.

File: api/app/Http/Controllers/FilledJobFormController.php [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+            });
+
+        return view('filled-form.show',[
+            'fields' => $fields,
+            'form' => $filledForm,
+        ]);
+    }
+
+    /**
+     * Return array of fields that ahs to be validated and their rules
```
Comment (EdotJ) [inline api/app/Http/Controllers/FilledJobFormController.php:116]: has*

File: api/resources/views/filled-form/edit.blade.php (hunk 13 above)
Comment (EdotJ) [reply api/resources/views/filled-form/edit.blade.php:13]: So if I insert JOE MAMA that's a legit response? Okay.

File: api/public/js/main.js (hunk 14 above)
Comment (ArvApu) [reply api/public/js/main.js:1]: oh right, will fix but probs not today

**Review by EdotJ — APPROVED** [review]

**Review by erteriuke — APPROVED** [review]
